test/unit/test_qpu.py

```diff
@@ -21,3 +54,48 @@ def test_provided_execution_options():
 
     qpu = QPU(quantum_processor_id="test", execution_options=options)
     assert qpu.execution_options == options
+
+
+@patch("pyquil.api._qpu.retrieve_results")
+@patch("pyquil.api._qpu.submit")
+def test_qpu_execute(mock_submit, mock_retrieve_results, mock_encrypted_program):
+    qpu = QPU(quantum_processor_id="test")
+
+    mock_submit.return_value = "some-job-id"
+    execute_response = qpu.execute(mock_encrypted_program)
+
+    mock_retrieve_results.return_value = ExecutionResults(
+        {
+            "q0": ExecutionResult.from_register(Register.from_i32([1, 1, 1, 1])),
+            "q1": ExecutionResult.from_register(Register.from_i32([1, 1, 1, 1])),
+        }
+    )
+
+    result = qpu.get_result(execute_response)
+
+    assert np.all(result.register_map["ro"] == np.array([[1, 1], [1, 1], [1, 1], [1, 1]]))
+    assert np.all(result.register_map["ro"] == result.readout_data["ro"])
+
+
+@patch("pyquil.api._qpu.retrieve_results")
+@patch("pyquil.api._qpu.submit")
+def test_qpu_execute_jagged_results(mock_submit, mock_retrieve_results, mock_encrypted_program):
```

Comment:
Please also test "not jagged but not of shots x memory shape"

Reply:
That will look equivalent to the `test_qpu_execute` test. Result data in the QCS SDK isn't aware of program metadata so converting to a register map will succeed as long as it fits a rectangular matrix, since the result could be valid for some number of shots. I conflated what users should do with the error case in some of the documentation, so I've made that more clear with a callout.